File: frontend/src/api/outboundLog.js
// Aussenkontakte-Aufzeichnung: Mapper + REST-Aufrufe (CERNIS PRO 2.0)
//
// Lese-/Schreibpfade der Aussenkontakte-Aufzeichnung (E5). Uebersetzt die
// Backend-Antworten (snake_case, aus backend/api/outbound_log.py) in die
// camelCase-View-Struktur. Stil bewusst wie api/monitoring.js: kleine reine
// Mapper, null = ehrlich nicht vorhanden (KEIN erfundener Fallback), Fehler ueber
// ApiError. Pfade bleiben relativ; NIE einen Host hartkodieren.
//
// Wire-Form (verifiziert, NICHT aendern), Prefix /api/outbound/recordings:
//   GET    ""                  -> [ {id,label,purpose,mode,depth,state,interval_s,
//                                    created_at,effective_start,max_duration_s} ]
//   POST   ""  Body {label,purpose?,mode,depth,interval_s?,max_duration_s?} -> 201 dict
//   GET    "/{id}"             -> dict | 404
//   POST   "/{id}/start"       -> dict | 404 | 409 (nur EINE darf host-weit active sein)
//   POST   "/{id}/pause"       -> dict | 404 | 409
//   POST   "/{id}/resume"      -> dict | 404 | 409
//   POST   "/{id}/stop"        -> dict | 404 | 409
//   DELETE "/{id}"             -> 204 (idempotent)
//   GET    "/{id}/aggregate"   -> [ {remote_ip,first_seen,last_seen,total_count,
//                                    peak_count,remote_port,hostname,country,operator,
//                                    asn,app_name} ]  (E5b)
//   GET    "/{id}/detail?since=&until=" -> [ {ts,remote_ip,remote_port,hostname,country,
//                                    operator,asn,app_name,pid,connection_count} ]  (E5b)
//
//   mode:  "detail" | "aggregate"
//   depth: "anonymous" | "app_resolved"
//   state: "created" | "active" | "paused" | "finished"

import { apiGet, apiPost, ApiError } from "./client.js";

// Basis-Pfad der Aufzeichnungs-Ressource (relativ; Vite-Proxy leitet ans Backend).
const BASIS = "/api/outbound/recordings";

// Lokaler DELETE-Helfer: client.js kennt nur GET/POST/PUT, kein apiDelete. Statt
// das Fundament fuer einen einzigen Aufruf zu erweitern, hier ein kleiner
// fetch-Helfer im EXAKTEN Stil von apiGet/apiPost (Muster api/monitoring.js) --
// gleiche ApiError-Form (status = HTTP-Code oder null bei Netzfehler). Bei 204
// (DELETE-Antwort ohne Body) wird null geliefert; sonst das geparste JSON.
async function apiDelete(path) {
  let response;
  try {
    response = await fetch(path, {
      method: "DELETE",
      headers: { Accept: "application/json" },
    });
  } catch (ursache) {
    // Netzfehler (Server nicht erreichbar, DNS, Abbruch o. Ae.): kein HTTP-Status.
    throw new ApiError(ursache?.message ?? "Netzwerkfehler", null);
  }

  if (!response.ok) {
    throw new ApiError(`Unerwarteter HTTP-Status ${response.status}`, response.status);
  }

  // 204 (No Content) traegt keinen Body -- response.json() wuerde werfen.
  if (response.status === 204) {
    return null;
  }
  return response.json();
}

// Lokaler POST-Helfer fuer die Lifecycle-Aktionen (start/pause/resume/stop). Wie
// apiPost, ABER er reicht bei einem Fehler den Backend-detail-Text durch und
// erhaelt den HTTP-Status (status) in der ApiError. Bei 409 (host-weit schon eine
// active / unzulaessige Transition) unterscheidet die View den ruhigen Konflikt-
// Hinweis (status === 409) von anderen Fehlern. Der generische apiPost aus
// client.js verwirft den detail-Text -- darum hier ein eigener Pfad, OHNE client.js
// umzubauen (Muster api/monitoring.js apiPostMitDetail). Kein Body (die id traegt
// der Pfad). Bei ok -> geparstes JSON; bei !ok -> ApiError (message = detail-Text
// bzw. Fallback, status = HTTP-Code).
async function apiPostMitDetail(path) {
  let response;
  try {
    response = await fetch(path, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
    });
  } catch (ursache) {
    // Netzfehler (Server nicht erreichbar, DNS, Abbruch o. Ae.): kein HTTP-Status.
    throw new ApiError(ursache?.message ?? "Netzwerkfehler", null);
  }

  if (!response.ok) {
    // detail aus dem Fehler-Body ziehen (FastAPI: {"detail": "..."}). Schlaegt das
    // Parsen fehl oder fehlt detail, faellt es auf eine generische Meldung zurueck.
    let detail = null;
    try {
      const body = await response.json();
      if (body && typeof body.detail === "string") {
        detail = body.detail;
      }
    } catch {
      // Kein/kein JSON-Body: detail bleibt null -> generische Meldung.
    }
    throw new ApiError(detail ?? `Unerwarteter HTTP-Status ${response.status}`, response.status);
  }

  return response.json();
}

// ── Mapper (Wire-dict -> View-Objekt) ───────────────────────────────────────

// Ein Aufzeichnungs-Wire-dict (snake_case) -> View-Objekt (camelCase). Zahlen/null
// bleiben roh: intervalS/maxDurationS/effectiveStart sind je nach Modus/Zustand
// null (z. B. maxDurationS bei aggregate; effectiveStart vor dem ersten Start) --
// das bleibt EHRLICH null, kein erfundener Default. Die Modus-/Tiefe-/Zustands-
// Strings tragen das Backend-Vokabular unveraendert.
export function mappeRecording(r) {
  return {
    id: r.id,
    label: r.label,
    purpose: r.purpose ?? null,
    mode: r.mode,
    depth: r.depth,
    state: r.state,
    intervalS: r.interval_s ?? null,
    createdAt: r.created_at ?? null,
    effectiveStart: r.effective_start ?? null,
    maxDurationS: r.max_duration_s ?? null,
  };
}

// Ein Aggregat-Wire-dict (E5b) -> View-Objekt. Je Gegenstelle eine verdichtete
// Zeile (erstmals/zuletzt/wie oft). Optionale Felder bleiben ehrlich null.
export function mappeAggregat(a) {
  return {
    remoteIp: a.remote_ip,
    firstSeen: a.first_seen ?? null,
    lastSeen: a.last_seen ?? null,
    totalCount: a.total_count,
    peakCount: a.peak_count,
    remotePort: a.remote_port ?? null,
    hostname: a.hostname ?? null,
    country: a.country ?? null,
    operator: a.operator ?? null,
    asn: a.asn ?? null,
    appName: a.app_name ?? null,
  };
}

// Ein Detail-Wire-dict (E5b) -> View-Objekt. Je Zeitpunkt eine Zeile (Momentaufnahme
// einer Gegenstelle). ts roh durchgereicht (die View formatiert selbst); optionale
// Felder bleiben ehrlich null.
export function mappeDetail(d) {
  return {
    ts: d.ts,
    remoteIp: d.remote_ip,
    remotePort: d.remote_port ?? null,
    hostname: d.hostname ?? null,
    country: d.country ?? null,
    operator: d.operator ?? null,
    asn: d.asn ?? null,
    appName: d.app_name ?? null,
    pid: d.pid ?? null,
    connectionCount: d.connection_count,
  };
}

// ── REST-Aufrufe ─────────────────────────────────────────────────────────────

// GET "" -> alle Aufzeichnungen (Wire-Form -> View). Leer -> [].
export async function fetchRecordings() {
  const backend = await apiGet(BASIS);
  return (backend ?? []).map(mappeRecording);
}

// POST "" -> 201 + angelegte Aufzeichnung (Zustand "created"). Der Body traegt nur
// die gesetzten Felder: purpose/intervalS/maxDurationS werden NUR mitgegeben, wenn
// definiert -- so greift sonst der Backend-Default (z. B. max_duration_s = 86400 bei
// detail, interval_s = 60). NIE null senden, wenn der Default greifen soll. id/state/
// createdAt setzt das Backend. apiPost wirft bei 422 (Validierung) eine ApiError.
export async function createRecording({
  label,
  purpose,
  mode,
  depth,
  intervalS,
  maxDurationS,
}) {
  const payload = { label, mode, depth };
  if (purpose !== undefined && purpose !== null) {
    payload.purpose = purpose;
  }
  if (intervalS !== undefined && intervalS !== null) {
    payload.interval_s = intervalS;
  }
  if (maxDurationS !== undefined && maxDurationS !== null) {
    payload.max_duration_s = maxDurationS;
  }
  const backend = await apiPost(BASIS, payload);
  return mappeRecording(backend);
}

// GET "/{id}" -> EINE Aufzeichnung. 404 (unbekannte id) -> ApiError.
export async function fetchRecording(id) {
  const backend = await apiGet(`${BASIS}/${id}`);
  return mappeRecording(backend);
}

// POST "/{id}/start" -> Aufzeichnung ("created"/"paused" -> "active"). Bei 409
// (host-weit schon eine active / unzulaessige Transition) wirft apiPostMitDetail
// eine ApiError mit status === 409 und dem Backend-detail-Text -- die View zeigt
// dann den ruhigen Konflikt-Hinweis.
export async function startRecording(id) {
  const backend = await apiPostMitDetail(`${BASIS}/${id}/start`);
  return mappeRecording(backend);
}

// POST "/{id}/pause" -> Aufzeichnung ("active" -> "paused"). 409 -> ApiError.
export async function pauseRecording(id) {
  const backend = await apiPostMitDetail(`${BASIS}/${id}/pause`);
  return mappeRecording(backend);
}

// POST "/{id}/resume" -> Aufzeichnung ("paused" -> "active"). 409 (host-weit schon
// eine active / unzulaessige Transition) -> ApiError mit status === 409.
export async function resumeRecording(id) {
  const backend = await apiPostMitDetail(`${BASIS}/${id}/resume`);
  return mappeRecording(backend);
}

// POST "/{id}/stop" -> Aufzeichnung ({"active","paused"} -> "finished"). 409 -> ApiError.
export async function stopRecording(id) {
  const backend = await apiPostMitDetail(`${BASIS}/${id}/stop`);
  return mappeRecording(backend);
}

// DELETE "/{id}" -> 204 (kein Body). Idempotent im Backend (unbekannte id ist kein
// Fehler). Liefert null (kein Wert zurueck).
export async function deleteRecording(id) {
  return apiDelete(`${BASIS}/${id}`);
}

// GET "/{id}/aggregate" -> verdichtete Gegenstellen-Liste (E5b). Wird in E5a noch
// nicht in der UI genutzt, aber schon hier angelegt. Leer -> [].
export async function fetchAggregate(id) {
  const backend = await apiGet(`${BASIS}/${id}/aggregate`);
  return (backend ?? []).map(mappeAggregat);
}

// GET "/{id}/detail?since=&until=" -> Zeitpunkt-Liste (E5b). since/until (Unix-ts,
// optional) schraenken den Zeitraum ein -- nur gesetzte Grenzen werden als Query-
// Param angehaengt (Muster api/monitoring.js). Wird in E5a noch nicht in der UI
// genutzt. Leer -> [].
export async function fetchDetail(id, since = null, until = null) {
  const params = {};
  if (since !== null && since !== undefined) {
    params.since = since;
  }
  if (until !== null && until !== undefined) {
    params.until = until;
  }
  const backend = await apiGet(`${BASIS}/${id}/detail`, params);
  return (backend ?? []).map(mappeDetail);
}

export default {
  mappeRecording,
  mappeAggregat,
  mappeDetail,
  fetchRecordings,
  createRecording,
  fetchRecording,
  startRecording,
  pauseRecording,
  resumeRecording,
  stopRecording,
  deleteRecording,
  fetchAggregate,
  fetchDetail,
};

File: backend/api/outbound_log.py
"""FastAPI-Router der Aussenkontakte-Aufzeichnung (E4) -- die REST-Lifecycle-/Lese-Pfade.

Aeusserer Ring: nimmt HTTP entgegen und ruft ausschliesslich Use-Cases aus
``application/`` (import-linter: api -> nur application). Konkrete Adapter, Ports und
``domain``-Typen werden hier NICHT importiert -- die Use-Cases kommen per
FastAPI-Dependency herein (Verdrahtung im Composition Root ``app.py``), und
Domaenen-Objekte werden ueber Attribut-Zugriff zu JSON serialisiert (Typ ``Any``, exakt
wie der monitoring-Router das haelt).

EIGENER Pfad-Zweig ``/api/outbound/recordings`` -- getrennt vom bestehenden
``/api/outbound/contacts`` (``api/outbound.py``), der die Live-Kontaktsicht liefert. Hier
geht es um die AUFZEICHNUNGS-DEFINITIONEN (Anlegen + Lebenszyklus) und ihren verdichteten
bzw. rohen Lesebefund.

``mode``/``depth`` reicht der Router als ROHE ``str`` durch -- die ``str`` -> Enum-Hebung
passiert NICHT hier, sondern autoritativ im Use-Case (``CreateOutboundRecording``, Muster
``capture_mode`` in ``CreateLoggingTask``); so importiert der api-Ring KEINE
Domaenen-Enums (CI-harter import-linter-Contract). Ein Fehlwert wirft im Use-Case einen
``ValueError`` (StrEnum-Konstruktor / Domaenen-``__post_init__``), den der Create-Endpunkt
auf 422 mappt.

Fehler-Mapping (gespiegelt von den monitoring-Logging-Endpunkten, ``api/monitoring.py``):
``RecordingNotFound`` -> 404, ``RecordingConflict`` -> 409 (mit der ``running_id`` im
Bezug ueber ``str(exc)``), ``InvalidRecordingTransition`` -> 409 (falscher
Ausgangszustand). ``InvalidRecordingTransition`` wird im monitoring-Bestand als
Domaenen-Exception ueber den application-Ring RE-EXPORTIERT und dort DIREKT gefangen
(``api/monitoring.py``: ``from application.monitoring import InvalidTaskTransition`` +
``except (LoggingTaskConflict, InvalidTaskTransition)``) -- exakt diesen Weg spiegeln wir
hier (``from application.outbound_log import InvalidRecordingTransition``), darum bleibt
der import-linter-Contract gewahrt (der Name kommt aus ``application``, nicht aus
``domain``). Zusaetzlich beim Create: ``ValueError`` (Enum-Hebung / Domaenen-Invarianten:
ungueltiger mode/depth/interval_s/DETAIL-Deckel) -> 422.
"""

import time
import uuid
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from application.outbound_log import (
    CreateOutboundRecording,
    DeleteOutboundRecording,
    EditOutboundRecording,
    GetOutboundAggregate,
    GetOutboundDetailRange,
    GetOutboundRecording,
    InvalidRecordingTransition,
    ListOutboundRecordings,
    PauseOutboundRecording,
    RecordingConfigLocked,
    RecordingConflict,
    RecordingNotFound,
    ResumeOutboundRecording,
    StartOutboundRecording,
    StopOutboundRecording,
)

router = APIRouter(prefix="/api/outbound/recordings", tags=["outbound_log"])


# ── Provider-Marker (im Composition Root ``app.py`` per dependency_overrides ersetzt) ──
# Jeder Marker wirft im ungebundenen Zustand -- keine stillen Fallbacks (Finding S3).


def provide_create_outbound_recording() -> CreateOutboundRecording:
    raise NotImplementedError("CreateOutboundRecording wird in app.py verdrahtet")


def provide_start_outbound_recording() -> StartOutboundRecording:
    raise NotImplementedError("StartOutboundRecording wird in app.py verdrahtet")


def provide_pause_outbound_recording() -> PauseOutboundRecording:
    raise NotImplementedError("PauseOutboundRecording wird in app.py verdrahtet")


def provide_resume_outbound_recording() -> ResumeOutboundRecording:
    raise NotImplementedError("ResumeOutboundRecording wird in app.py verdrahtet")


def provide_stop_outbound_recording() -> StopOutboundRecording:
    raise NotImplementedError("StopOutboundRecording wird in app.py verdrahtet")


def provide_edit_outbound_recording() -> EditOutboundRecording:
    raise NotImplementedError("EditOutboundRecording wird in app.py verdrahtet")


def provide_delete_outbound_recording() -> DeleteOutboundRecording:
    raise NotImplementedError("DeleteOutboundRecording wird in app.py verdrahtet")


def provide_list_outbound_recordings() -> ListOutboundRecordings:
    raise NotImplementedError("ListOutboundRecordings wird in app.py verdrahtet")


def provide_get_outbound_recording() -> GetOutboundRecording:
    raise NotImplementedError("GetOutboundRecording wird in app.py verdrahtet")


def provide_get_outbound_aggregate() -> GetOutboundAggregate:
    raise NotImplementedError("GetOutboundAggregate wird in app.py verdrahtet")


def provide_get_outbound_detail_range() -> GetOutboundDetailRange:
    raise NotImplementedError("GetOutboundDetailRange wird in app.py verdrahtet")


# ── Wire-Modelle (eigene Wire-Form, KEINE application/domain-Typen) ───────────


class CreateRecordingBody(BaseModel):
    """Anlage-Body einer Aufzeichnung. ``mode``/``depth`` als rohe ``str`` (Wire).

    ``mode``/``depth`` reicht der Router roh durch -- die Vokabular-Validierung +
    Enum-Hebung macht der Use-Case (422 bei Fehlwert). ``interval_s`` Default 60
    (Domaenen-Default), ``max_duration_s`` optional (nur DETAIL-relevant; AGGREGATE
    erzwingt der Use-Case ohnehin auf ``None``).
    """

    label: str
    purpose: str = ""
    mode: str
    depth: str
    interval_s: int = 60
    max_duration_s: int | None = None


class EditRecordingBody(BaseModel):
    """Aenderungs-Body einer Aufzeichnung -- gleiche Felder wie ``CreateRecordingBody``.

    ``label``/``purpose`` sind in jedem Zustand aenderbar; die Konfig-Felder
    ``mode``/``depth``/``interval_s``/``max_duration_s`` nur im Zustand CREATED (sonst
    409 via ``RecordingConfigLocked``). ``mode``/``depth`` als rohe ``str`` (Enum-Hebung
    im Use-Case, 422 bei Fehlwert).
    """

    label: str
    purpose: str = ""
    mode: str
    depth: str
    interval_s: int = 60
    max_duration_s: int | None = None


# ── Serialisierungs-Helfer (Domaenen-Objekt -> Wire-dict am api-Rand) ─────────


def _recording_to_dict(rec: Any) -> dict[str, Any]:
    """``OutboundRecording`` -> Wire-dict. Die StrEnums (``mode``/``depth``/``state``)
    als ihr ``str``-Wert (``str(...)`` ergibt den Vokabular-String); die Zeitfelder roh
    durchgereicht (Unix-ts bzw. ``None``).
    """
    return {
        "id": rec.id,
        "label": rec.label,
        "purpose": rec.purpose,
        "mode": str(rec.mode),
        "depth": str(rec.depth),
        "state": str(rec.state),
        "interval_s": rec.interval_s,
        "created_at": rec.created_at,
        "effective_start": rec.effective_start,
        "max_duration_s": rec.max_duration_s,
    }


def _aggregate_to_dict(c: Any) -> dict[str, Any]:
    """``AggregatedContact`` -> Wire-dict. Alle Felder roh durchgereicht (Werte oder ``None``)."""
    return {
        "remote_ip": c.remote_ip,
        "first_seen": c.first_seen,
        "last_seen": c.last_seen,
        "total_count": c.total_count,
        "peak_count": c.peak_count,
        "remote_port": c.remote_port,
        "hostname": c.hostname,
        "country": c.country,
        "operator": c.operator,
        "asn": c.asn,
        "app_name": c.app_name,
    }


def _detail_to_dict(r: Any) -> dict[str, Any]:
    """``OutboundDetailRow`` -> Wire-dict. Alle Felder roh durchgereicht (Werte oder ``None``)."""
    return {
        "ts": r.ts,
        "remote_ip": r.remote_ip,
        "remote_port": r.remote_port,
        "hostname": r.hostname,
        "country": r.country,
        "operator": r.operator,
        "asn": r.asn,
        "app_name": r.app_name,
        "pid": r.pid,
        "connection_count": r.connection_count,
    }


# ── Lifecycle-/Lese-Endpunkte ────────────────────────────────────────────────


@router.post("", status_code=status.HTTP_201_CREATED)
def create_recording(
    body: CreateRecordingBody,
    create_recording_uc: Annotated[
        CreateOutboundRecording, Depends(provide_create_outbound_recording)
    ],
) -> dict[str, Any]:
    """Legt eine Aufzeichnung an (Zustand CREATED) -> 201 mit der Wire-Form.

    Der Router erzeugt ``id`` (uuid4-hex) + ``now`` (time.time()) und reicht ``mode``/
    ``depth`` roh durch; die Enum-Hebung + Wert-Invarianten (interval_s/DETAIL-Deckel)
    pruefen Use-Case/Domaene -- ein ``ValueError`` daraus -> 422 (kein 500/stiller Fallback).
    """
    try:
        rec = create_recording_uc(
            recording_id=uuid.uuid4().hex,
            label=body.label,
            purpose=body.purpose,
            mode=body.mode,
            depth=body.depth,
            interval_s=body.interval_s,
            now=time.time(),
            max_duration_s=body.max_duration_s,
        )
    except ValueError as exc:
        # 422 als nacktes Literal (Bestandsmuster ``api/monitoring.py``); das
        # ``status.HTTP_422_*``-Symbol ist in starlette inzwischen deprecation-markiert.
        raise HTTPException(422, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.get("")
def list_recordings(
    list_uc: Annotated[ListOutboundRecordings, Depends(provide_list_outbound_recordings)],
) -> list[dict[str, Any]]:
    """Alle Aufzeichnungs-Definitionen als Wire-Form. Leer -> ``[]``."""
    return [_recording_to_dict(r) for r in list_uc()]


@router.get("/{recording_id}")
def get_recording(
    recording_id: str,
    get_uc: Annotated[GetOutboundRecording, Depends(provide_get_outbound_recording)],
) -> dict[str, Any]:
    """EINE Aufzeichnungs-Definition. 404 bei unbekannter id."""
    try:
        rec = get_uc(recording_id)
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.post("/{recording_id}/start")
def start_recording(
    recording_id: str,
    start_uc: Annotated[StartOutboundRecording, Depends(provide_start_outbound_recording)],
) -> dict[str, Any]:
    """Startet eine Aufzeichnung (CREATED -> ACTIVE). 404/409 bei Fehler.

    ``RecordingConflict`` -> 409 (host-weit nur eine aktiv, ``running_id`` im Bezug);
    ``InvalidRecordingTransition`` -> 409 (falscher Ausgangszustand).
    """
    try:
        rec = start_uc(recording_id, time.time())
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except (RecordingConflict, InvalidRecordingTransition) as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.post("/{recording_id}/pause")
def pause_recording(
    recording_id: str,
    pause_uc: Annotated[PauseOutboundRecording, Depends(provide_pause_outbound_recording)],
) -> dict[str, Any]:
    """Pausiert eine Aufzeichnung (ACTIVE -> PAUSED). 404/409 (InvalidRecordingTransition)."""
    try:
        rec = pause_uc(recording_id)
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidRecordingTransition as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.post("/{recording_id}/resume")
def resume_recording(
    recording_id: str,
    resume_uc: Annotated[ResumeOutboundRecording, Depends(provide_resume_outbound_recording)],
) -> dict[str, Any]:
    """Setzt eine Aufzeichnung fort (PAUSED -> ACTIVE). 404/409 (Konflikt/Transition).

    Fortsetzen aus Pause ist im Konzept ein Start (host-weit nur eine aktiv) -> ein
    ``RecordingConflict`` ist hier ebenso moeglich wie beim ``start``.
    """
    try:
        rec = resume_uc(recording_id)
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except (RecordingConflict, InvalidRecordingTransition) as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.post("/{recording_id}/stop")
def stop_recording(
    recording_id: str,
    stop_uc: Annotated[StopOutboundRecording, Depends(provide_stop_outbound_recording)],
) -> dict[str, Any]:
    """Beendet eine Aufzeichnung ({ACTIVE,PAUSED} -> FINISHED). 404/409 (Transition)."""
    try:
        rec = stop_uc(recording_id)
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidRecordingTransition as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.put("/{recording_id}")
def edit_recording(
    recording_id: str,
    body: EditRecordingBody,
    edit_uc: Annotated[EditOutboundRecording, Depends(provide_edit_outbound_recording)],
) -> dict[str, Any]:
    """Aendert eine Aufzeichnungs-DEFINITION. 404/409/422 bei Fehler.

    ``label``/``purpose`` sind in jedem Zustand aenderbar; die vier Konfig-Felder nur im
    Zustand CREATED. ``RecordingNotFound`` -> 404, ``RecordingConfigLocked`` -> 409
    (Konfig-Aenderung ausserhalb CREATED), ``ValueError`` -> 422 (Enum-Hebung /
    Domaenen-Invarianten: ungueltiger mode/depth/interval_s/DETAIL-Deckel).
    """
    try:
        rec = edit_uc(
            recording_id=recording_id,
            label=body.label,
            purpose=body.purpose,
            mode=body.mode,
            depth=body.depth,
            interval_s=body.interval_s,
            max_duration_s=body.max_duration_s,
        )
    except RecordingNotFound as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except RecordingConfigLocked as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    except ValueError as exc:
        # 422 als nacktes Literal (Bestandsmuster wie ``create_recording``).
        raise HTTPException(422, detail=str(exc)) from exc
    return _recording_to_dict(rec)


@router.delete("/{recording_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_recording(
    recording_id: str,
    delete_uc: Annotated[DeleteOutboundRecording, Depends(provide_delete_outbound_recording)],
) -> None:
    """Loescht eine Aufzeichnungs-DEFINITION. Idempotent (unbekannte id ist kein Fehler) -> 204."""
    delete_uc(recording_id)


@router.get("/{recording_id}/aggregate")
def get_aggregate(
    recording_id: str,
    get_aggregate_uc: Annotated[GetOutboundAggregate, Depends(provide_get_outbound_aggregate)],
) -> list[dict[str, Any]]:
    """Verdichteter Stand einer Aufzeichnung als Wire-Liste. Leer/unbekannte id -> ``[]``."""
    return [_aggregate_to_dict(c) for c in get_aggregate_uc(recording_id)]


@router.get("/{recording_id}/detail")
def get_detail(
    recording_id: str,
    get_detail_range_uc: Annotated[
        GetOutboundDetailRange, Depends(provide_get_outbound_detail_range)
    ],
    since: Annotated[float, Query()],
    until: Annotated[float, Query()],
) -> list[dict[str, Any]]:
    """Roher DETAIL-Zeitverlauf einer Aufzeichnung im Fenster ``[since, until)`` als Wire-Liste.

    ``since``/``until`` sind ABSOLUTE Query-Floats (Unix-ts) -- der Use-Case haelt keine
    Uhr, das Fenster kommt vom Client. Keine Daten / unbekannte id -> ``[]``.
    """
    return [_detail_to_dict(r) for r in get_detail_range_uc(recording_id, since, until)]
